File: Easy/977 - SquaresOfASortedArray/solution.ts
/**
 * Question 977. Squares of a Sorted Array (https://leetcode-cn.com/problems/squares-of-a-sorted-array/)
 *
 * 执行用时：144 ms, 在所有 typescript 提交中击败了 72.62% 的用户
 * 内存消耗：45.8 MB, 在所有 typescript 提交中击败了 5.00% 的用户
 *
 * @param {number[]} A
 * @returns {number[]}
 */
function sortedSquares(A: number[]): number[] {
  return A.map((a) => a * a).sort((a, b) => a - b);
}

/**
 * 执行用时：128 ms, 在所有 typescript 提交中击败了 92.86% 的用户
 * 内存消耗：46.7 MB, 在所有 typescript 提交中击败了 5.00% 的用户
 *
 * @param {number[]} A
 * @returns {number[]}
 */
function sortedSquaresTwoPointers(A: number[]): number[] {
  if (!A.length) {
    return A;
  }

  // This array contains no negative number.
  if (A[0] >= 0) {
    return A.map((a) => a * a);
  }

  const N: number = A.length;
  // This array contains no positive number.
  if (A[N - 1] < 0) {
    const sorted: number[] = [];

    for (let i = N - 1; i >= 0; i--) {
      const currentValue: number = A[i];
      sorted.push(currentValue * currentValue);
    }

    return sorted;
  }

  let negativeIndex: number = 0;
  let positiveIndex: number = 0;
  let sorted: number[] = [];

  while (positiveIndex < N) {
    if (A[positiveIndex] >= 0) {
      negativeIndex = positiveIndex - 1;
      break;
    }

    positiveIndex++;
  }

  let positiveSquare: number = A[positiveIndex] * A[positiveIndex];
  let negativeSquare: number = A[negativeIndex] * A[negativeIndex];

  while (positiveIndex < N || negativeIndex >= 0) {
    // console.log(
    //   `positiveIndex = ${positiveIndex}, positiveSquare = ${positiveSquare}, negativeIndex = ${negativeIndex}, negativeSquare = ${negativeSquare}`
    // );

    if (positiveSquare < negativeSquare) {
      sorted.push(positiveSquare);
      positiveIndex++;

      if (positiveIndex < N) {
        const currentValue: number = A[positiveIndex];
        positiveSquare = currentValue * currentValue;
      } else {
        positiveSquare = Number.MAX_SAFE_INTEGER;
      }
    } else if (positiveSquare === negativeSquare) {
      sorted.push(positiveSquare);
      positiveIndex++;

      if (positiveIndex < N) {
        const currentValue: number = A[positiveIndex];
        positiveSquare = currentValue * currentValue;
      } else {
        positiveSquare = Number.MAX_SAFE_INTEGER;
      }

      sorted.push(negativeSquare);
      negativeIndex--;

      if (negativeIndex >= 0) {
        const currentValue: number = A[negativeIndex];
        negativeSquare = currentValue * currentValue;
      } else {
        negativeSquare = Number.MAX_SAFE_INTEGER;
      }
    } else {
      sorted.push(negativeSquare);
      negativeIndex--;

      if (negativeIndex >= 0) {
        const currentValue: number = A[negativeIndex];
        negativeSquare = currentValue * currentValue;
      } else {
        negativeSquare = Number.MAX_SAFE_INTEGER;
      }
    }
  }

  return sorted;
}

File: Easy/977 - SquaresOfASortedArray/Solution.cs
/// <summary>
/// Question 977. Squares of a Sorted Array (https://leetcode-cn.com/problems/squares-of-a-sorted-array/)
/// </summary>

public class Solution
{
    /// <summary>
    /// 执行用时：340 ms, 在所有 C# 提交中击败了 97.01% 的用户
    /// 内存消耗：39.9 MB, 在所有 C# 提交中击败了 28.81% 的用户
    /// 
    /// Acknowledgement: This solution borrows idea from the third method of the [official solution](https://leetcode-cn.com/problems/squares-of-a-sorted-array/solution/you-xu-shu-zu-de-ping-fang-by-leetcode-solution/)
    /// </summary>
    /// <param name="A"></param>
    /// <returns></returns>
    public int[] SortedSquares(int[] A)
    {
        int length = A.Count();

        if (length == 0) { return A; }

        int[] sorted = new int[length];
        int left = 0;
        int right = length - 1;
        int insertIndex = length - 1;
        int leftValue = A[left] * A[left];
        int rightValue = A[right] * A[right];

        while (left <= right)
        {
            if (rightValue >= leftValue)
            {
                sorted[insertIndex] = rightValue;
                insertIndex--;
                right--;
                if (right >= 0) { rightValue = A[right] * A[right]; }
            }
            else if (rightValue < leftValue)
            {
                sorted[insertIndex] = leftValue;
                insertIndex--;
                left++;

                if (left < length) { leftValue = A[left] * A[left]; }
            }

        }

        return sorted;
    }
}